feat(aio): Convert AIO to use the new Service Worker 5.0.0.

alxhub opened the pull request:
AIO is currently using a beta version of @angular/service-worker.
Since that was implemented, the SW has been rewritten and released
as part of Angular 5.0.0. This commit updates AIO to use the latest
implementation, with an appropriate configuration file that caches
the various AIO assets in useful ways.

gkalpak commented on aio/src/main.ts:
```diff
@@ -8,11 +8,4 @@ if (environment.production) {
 }
 
-platformBrowserDynamic().bootstrapModule(AppModule).then(ref => {
-  if (environment.production && 'serviceWorker' in (navigator as any)) {
```
Will the ServiceWorker still be registered on production only?
(How does it know which environment it is in?)

alxhub replied:
The module is only added if `environment.production` is set.

IgorMinar replied:
Will that work fine if the SW support doesn't exist at runtime (e.g. ios)?

alxhub replied:
Yes, the module itself does feature detection to decide whether to register the SW.

gkalpak commented on aio/src/ngsw.json:
```diff
@@ -0,0 +1,50 @@
+      ],
+      "urls": [
+        "https://fonts.googleapis.com/**",
+        "http://fonts.gstatic.com/s/**",
```
Shouldn't this be `https` too?

gkalpak commented on aio/src/ngsw.json:
```diff
@@ -0,0 +1,50 @@
+{
+  "index": "/index.html",
```
How does the new ServiceWorker handle routing? (For reference this was our [previous config][1].)

[1]: https://github.com/alxhub/angular/blob/8c863e030260a115ea08c0db558280684b0f754d/aio/ngsw-manifest.json#L19-L26

gkalpak commented on aio/src/ngsw.json:
```diff
@@ -0,0 +1,50 @@
+        "/generated/docs/api/api-list.json"
+      ]
+    }
+  }]
```
AFAICT, it is possible to follow the [previous config][1] more closely (if not exactly) wrt to what gets cached and when.
Have you intentionally deviated from it? 

[1]: https://github.com/alxhub/angular/blob/8c863e030260a115ea08c0db558280684b0f754d/aio/ngsw-manifest.json

gkalpak reviewed (commented):
🎉 ✨ (Quite excited about the dynamic content caching 😄)
A couple of comments/questions:

- You can now remove the `--ignore-packages @angular/service-worker` part from [here][1].
- The [`SwUpdatesService` spec][2] needs to be updated too :grin:
- Do we still need [ngsw-manifest.json][3]?

[1]: https://github.com/alxhub/angular/blob/8c863e030260a115ea08c0db558280684b0f754d/aio/package.json#L9
[2]: https://github.com/alxhub/angular/blob/8c863e030260a115ea08c0db558280684b0f754d/aio/src/app/sw-updates/sw-updates.service.spec.ts
[3]: https://github.com/alxhub/angular/blob/8c863e030260a115ea08c0db558280684b0f754d/aio/ngsw-manifest.json

lacolaco commented on aio/package.json:
```diff
@@ -50,8 +50,8 @@
-    "sw-manifest": "ngu-sw-manifest --dist dist --in ngsw-manifest.json --out dist/ngsw-manifest.json",
-    "sw-copy": "cp node_modules/@angular/service-worker/bundles/worker-basic.min.js dist/",
+    "sw-manifest": "ngsw-config dist src/ngsw.json",
+    "sw-copy": "cp node_modules/@angular/service-worker/ngsw-worker.js dist/",
```
Just out of curiosity, couldn't you use `assets` of the CLI?

alxhub replied:
The SW script has to be registered at the root of the page's origin, it can't come from a subdirectory.

webmaxru commented on aio/src/ngsw.json:
```diff
@@ -0,0 +1,50 @@
+  "index": "/index.html",
+  "assetGroups": [{
+    "name": "site",
+    "mode": "prefetch",
```
Do you mean `installMode`?

alxhub replied:
Fixed.

webmaxru commented on aio/src/ngsw.json:
```diff
@@ -0,0 +1,50 @@
+        "/pwa-manifest.json",
+        "/app/search/search-worker.js"
+      ],
+      "urls": [
```
I'd move this block with fonts to a separate group, because it will not follow `prefetch` rule anyway

alxhub replied:
It's conceptually part of the site.

webmaxru commented on aio/src/app/sw-updates/sw-updates.service.ts:
```diff
@@ -31,38 +30,27 @@ export class SwUpdatesService implements OnDestroy {
-  updateActivated = this.sw.updates
+  updateActivated = this.sw.activated
       .takeUntil(this.onDestroy)
       .do(evt => this.log(`Update event: ${JSON.stringify(evt)}`))
```
My IDE says we have to add `import 'rxjs/add/operator/do';`

alxhub replied:
Do(ne).

Splaktar commented on aio/package.json:
```diff
@@ -77,7 +77,7 @@
     "@angular/platform-server": "^5.0.0-beta.3",
     "@angular/router": "^5.0.0-beta.3",
-    "@angular/service-worker": "^1.0.0-beta.16",
+    "@angular/service-worker": "5.0.0-rc.2",
```
Need to rebase to master to pick up https://github.com/angular/angular/commit/6b748835be3240a4e26fcc3b6f9829753490736a and then update this to `5.0.0`. It probably also worth using the Angular CLI `1.6.0-beta.0` release that supports Angular Service Worker.

Splaktar reviewed (commented):
This will be a really useful 'real world' example once it gets to production!

IgorMinar commented:
@alxhub can you please resurrect this PR and get it into a mergable state? thanks!

alxhub committed:
feat(docs-infra): Convert AIO to use the new Service Worker 5.0.0.

AIO is currently using a beta version of @angular/service-worker.
Since that was implemented, the SW has been rewritten and released
as part of Angular 5.0.0. This commit updates AIO to use the latest
implementation, with an appropriate configuration file that caches
the various AIO assets in useful ways.

alxhub commented:
@IgorMinar PTAL.

IgorMinar commented on aio/scripts/_payload-limits.json:
```diff
@@ -1,3 +1,3 @@
 {
-"aio":{"master":{"change":"application","gzip7":{"inline":925,"main":119519,"polyfills":11863},"gzip9":{"inline":925,"main":119301,"polyfills":11861},"uncompressed":{"inline":1533,"main":486493,"polyfills":37068}}}
+"aio":{"master":{"change":"application","gzip7":{"inline":925,"main":119519,"polyfills":11863},"gzip9":{"inline":925,"main":119301,"polyfills":11861},"uncompressed":{"inline":1533,"main":486493,"polyfills":37070}}}
```
Also, I'd expect the main size to decrease. Isn't that the case or is it less than 1% so the size check in CI doesn't fail?

alxhub replied:
I'm not sure what you mean. It's probably less than 1%, this doesn't change much code.

IgorMinar commented on aio/src/app/app.module.ts:
```diff
@@ -89,7 +92,8 @@ export const svgIconProviders = [
     SwUpdatesModule,
-    SharedModule
+    SharedModule,
+    environment.production ? ServiceWorkerModule.register('/ngsw-worker.js') : [],
```
this is a bit odd, because it means that now you have to conditionally or optionally inject the ngsw services which makes the application code not very ergonomic.

alxhub replied:
Good point. I should fix that.

IgorMinar commented on aio/src/app/sw-updates/sw-updates.service.ts:
```diff
@@ -31,18 +33,27 @@ export class SwUpdatesService implements OnDestroy {
+      this.updateActivated = Observable.never<string>().takeUntil(this.onDestroy);
+      return;
+    }
+    this.sw = injector.get(SwUpdate);
```
ouch. that's ugly. can we do better? either inject it optionally, or even better always register the providers and inject them, but if we are on browser without SW support or in Dev mode, then make the services inert.. or something like that. We should really rethink this, because the current ergonomics are bad. I believe that this is what Stephen was pointing out as well...

IgorMinar commented on aio/src/ngsw-config.json:
```diff
@@ -0,0 +1,50 @@
+{
+  "index": "/index.html",
+  "assetGroups": [{
+    "name": "site",
```
please rename this to "app-shell"

alxhub replied:
Done.

IgorMinar commented on aio/src/ngsw-config.json:
```diff
@@ -0,0 +1,50 @@
+      ]
+    }
+  }, {
+    "name": "docs",
```
there are actually two docs content groups - the ones that are available offline, and those that are lazy downloaded.

The offline docs include all marketing pages and images under generated/marketing/ I believe - check what the current site is fetching: https://angular.io/ngsw-manifest.json

The the remaining docs and assets should be downloaded on as needed basis. Your proposal changes the update behavior (compared to current angular.io) - I'm fine experimenting with the prefetch mode. We'll see how well this user experience works.

alxhub replied:
As discussed, the `docs` group has been moved to `prefetch`.

IgorMinar reviewed (dismissed):
I'm a bit concerned about the impact on ergonomics, let's discuss how we can resolve that - I'm afraid that this will require an API change/addition unless you have other ideas. Otherwise this looks good after the nitpicks are resolved.

alxhub commented on aio/package.json:
```diff
@@ -49,8 +49,8 @@
-    "sw-manifest": "ngu-sw-manifest --dist dist --in ngsw-manifest.json --out dist/ngsw-manifest.json",
-    "sw-copy": "cp node_modules/@angular/service-worker/bundles/worker-basic.min.js dist/",
+    "sw-manifest": "ngsw-config dist src/ngsw-config.json",
+    "sw-copy": "cp node_modules/@angular/service-worker/ngsw-worker.js dist/",
```
Yes, I could. But I don't want to make more changes to AIO's build system than I need to.

gkalpak replied:
FWIW, we should soon upgrade to the latest-ish cli in #18428. So, maybe after that is merged you can use the build integration.

alxhub commented on aio/scripts/_payload-limits.json:
```diff
@@ -1,3 +1,3 @@
 {
-"aio":{"master":{"change":"application","gzip7":{"inline":925,"main":119519,"polyfills":11863},"gzip9":{"inline":925,"main":119301,"polyfills":11861},"uncompressed":{"inline":1533,"main":486493,"polyfills":37068}}}
+"aio":{"master":{"change":"application","gzip7":{"inline":925,"main":119519,"polyfills":11863},"gzip9":{"inline":925,"main":119301,"polyfills":11861},"uncompressed":{"inline":1533,"main":486493,"polyfills":37070}}}
```
No idea - it was failing.

gkalpak replied:
It couldn't be failing for a 2B diff 😕
(FWIW, I think the polyfills size has been 37070 for a while, but it wasn't updated in the limits, since it is a small changed - much lower than 1%).

gkalpak commented on aio/src/app/app.module.ts:
```diff
@@ -53,6 +54,8 @@ import { WindowToken, windowProvider } from 'app/shared/window';
 
 import { SharedModule } from 'app/shared/shared.module';
 
+import {environment} from '../environments/environment';
```
Inconsistent spacing inside `{...}`.
<sub>I don't like the whitespace either, but... ¯\\\_(ツ)_/¯</sub>

alxhub replied:
Fixed.

gkalpak commented on aio/src/main.ts:
```diff
@@ -8,11 +8,4 @@ if (environment.production) {
-platformBrowserDynamic().bootstrapModule(AppModule).then(ref => {
-  if (environment.production && 'serviceWorker' in (navigator as any)) {
-    const appRef: ApplicationRef = ref.injector.get(ApplicationRef);
-    appRef.isStable.first().subscribe(() => {
```
IIRC, we added this to improve TTI or something.
Is this now handled by the `ServiceWorkerModule` internally?

alxhub replied:
Exactly.

mary-poppins commented:
You can preview 6509835 at https://pr19795-6509835.ngbuilds.io/.

alxhub commented on aio/src/index.html:
```diff
@@ -2,7 +2,7 @@
 <head>
   <meta charset="utf-8">
-  <title>Angular Docs</title>
+  <title>Angular Docs!!!</title>
```
Fixed.

IgorMinar replied:
It's not fixed. Did you forget to push the change?

mary-poppins commented:
You can preview 2a113a7 at https://pr19795-2a113a7.ngbuilds.io/.

mary-poppins commented:
You can preview 18b9846 at https://pr19795-18b9846.ngbuilds.io/.

mary-poppins commented:
You can preview 63aa878 at https://pr19795-63aa878.ngbuilds.io/.

mary-poppins commented:
You can preview 8bd1843 at https://pr19795-8bd1843.ngbuilds.io/.

mary-poppins commented:
You can preview 4e26eed at https://pr19795-4e26eed.ngbuilds.io/.

gkalpak commented on aio/src/ngsw-config.json:
```diff
@@ -0,0 +1,51 @@
+    "installMode": "prefetch",
+    "resources": {
+      "files": [
+        "/*.svg",
```
Do we really have top-level `.svg` files?

gkalpak commented on aio/src/ngsw-config.json:
```diff
@@ -0,0 +1,51 @@
+      "urls": [
+        "https://fonts.googleapis.com/**",
+        "https://fonts.gstatic.com/s/**",
+        "https://maxcdn.bootstrapcdn.com/**"
```
Does this mean that it will now wait for  request that matches a URL pattern first?
Does this change anything (in terms of TTI) compared to the previous implementation?

alxhub replied:
Yes, it will wait for a request that matches the pattern first.

No, once all files are cached the TTI is the same.

gkalpak commented on aio/src/ngsw-config.json:
```diff
@@ -0,0 +1,51 @@
+    "resources": {
+      "files": [
+        "/assets/images/favicons/**",
+        "/assets/images/logos/**",
```
Why not `/assets/images/**`?

alxhub replied:
This particular configuration is designed to match as closely as possible the behavior of the old SW. Once we validate and deploy it, I expect we will then look at expanding the set of cached assets.

gkalpak commented on aio/src/ngsw-config.json:
```diff
@@ -0,0 +1,51 @@
+      "files": [
+        "/assets/images/favicons/**",
+        "/assets/images/logos/**",
+        "/generated/images/marketing/**",
```
Wouldn't we want these to be eagerly fetched on install also? cc @IgorMinar

alxhub replied:
The idea behind fetching these lazily is that there are different sizes of e.g. the favicons, and this ensures you won't waste space in the cache for assets which aren't ever loaded.

On update, the ones which _did_ get loaded will be eagerly fetched.

gkalpak commented on aio/src/ngsw-config.json:
```diff
@@ -0,0 +1,51 @@
+        "/favicon.ico",
+        "/index.html",
+        "/pwa-manifest.json",
+        "/app/search/search-worker.js"
```
What about `/assets/js/**`?

gkalpak commented on aio/src/ngsw-config.json:
```diff
@@ -0,0 +1,51 @@
+        "/generated/docs/*.json",
+        "/generated/docs/api/api-list.json",
+        "/generated/docs/app/search-data.json"
+      ]
```
Do we really want to prefetch all of those? I assume they will be several MB. Will our users be happy with that?
Previouly, we only fetched `/generated/*` and `generated/docs/api/api-list.json`?

As discussed in the past, it would be awesome to have a way to request everything to be downloaded and be available offline - e.g. before your 3-hour flight :smiley: - but shouldn't it be an opt-in?
Or are we not that much concerned because the cost will mostly be paid once (given that guides don't change that often)?

alxhub replied:
The search-data is 504K (120K gzipped). Per discussing with Igor, we really the search index to be available in the cache ASAP.

gkalpak replied:
Ooops, I somehow thought this would fetch all `.json` files inside `generated/docs`. I now see it will only fetch top-level `.json` files inside `generated/docs/` (which is what we want).

alxhub commented on aio/src/app/sw-updates/sw-updates.service.spec.ts:
```diff
@@ -59,31 +58,22 @@ describe('SwUpdatesService', () => {
-    tick(checkInterval);
-
-    expect(sw.activateUpdate).toHaveBeenCalledWith(null);
+    sw.$$activatedSubj.next({current: {hash: 'bar'}});
```
Apparently not :)

alxhub commented on aio/src/app/sw-updates/sw-updates.service.spec.ts:
```diff
@@ -59,31 +58,22 @@ describe('SwUpdatesService', () => {
+    sw.$$activatedSubj.next({current: {hash: 'bar'}});
   })));
 
   it('should schedule a new check after activating the update', fakeAsync(run(() => {
```
Fixed with copy/paste :)

alxhub commented on aio/src/app/sw-updates/sw-updates.service.spec.ts:
```diff
@@ -145,16 +135,20 @@ class MockLogger {
+    return true;
+  }
 
   activateUpdate = jasmine.createSpy('MockNgServiceWorker.activateUpdate')
```
Done.

alxhub commented on aio/src/app/sw-updates/sw-updates.service.spec.ts:
```diff
@@ -145,16 +135,20 @@ class MockLogger {
-                          .and.callFake(() => this.$$activateUpdateSubj.take(1));
+                          .and.callFake(() => Promise.resolve());
 
   checkForUpdate = jasmine.createSpy('MockNgServiceWorker.checkForUpdate')
```
And done.

alxhub commented on aio/src/app/sw-updates/sw-updates.service.ts:
```diff
@@ -1,10 +1,12 @@
+import { SwUpdate } from '@angular/service-worker';
 import { Observable } from 'rxjs/Observable';
 import { Subject } from 'rxjs/Subject';
 import 'rxjs/add/observable/of';
```
Gone.

alxhub commented on aio/src/app/sw-updates/sw-updates.service.ts:
```diff
@@ -1,10 +1,12 @@
 import 'rxjs/add/operator/concat';
 import 'rxjs/add/operator/debounceTime';
+import 'rxjs/add/operator/do';
 import 'rxjs/add/operator/filter';
```
Gone.

alxhub commented on aio/src/app/sw-updates/sw-updates.service.ts:
```diff
@@ -31,18 +33,28 @@ export class SwUpdatesService implements OnDestroy {
-        .subscribe(() => this.checkForUpdate());
+        .subscribe(() => {
+          this.checkForUpdate();
+          this.checkForUpdateSubj.next();
```
Why not? It's a low cost operation and there's always a chance that a new update is pushed before the old update is activated.

mary-poppins commented:
You can preview 5adc026 at https://pr19795-5adc026.ngbuilds.io/.

mary-poppins commented:
You can preview 5cc0fba at https://pr19795-5cc0fba.ngbuilds.io/.

IgorMinar commented:
@alxhub can you please take a look at #21318 and ensure that the new SW is configured correctly to enable the firebase redirects to work correctly. Please note that the preview server does not honor the firebase config so AFAIK we can't test this on the preview server.

alxhub commented:
@IgorMinar there is no way in the new Service Worker to ignore routes - the SW will always serve the application on a navigation request.

gkalpak commented:
> there is no way in the new Service Worker to ignore routes - the SW will always serve the application on a navigation request.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;😱&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;😱&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;😱&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;😱&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;😱&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;😱&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;😱&nbsp;&nbsp;&nbsp;&nbsp;😱&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;😱&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;😱&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;😱&nbsp;&nbsp;
&nbsp;😱&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;😱&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;😱😱&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;😱&nbsp;&nbsp;&nbsp;😱😱&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;😱&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;😱&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;😱&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;😱&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;😱&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;😱&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;😱&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;😱&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;😱&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;😱&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;

gkalpak commented:
> there is no way in the new Service Worker to ignore routes - the SW will always serve the application on a navigation request.

Thinking about this again, does this mean that files that are not included in the assets will be forwarded to the server (which means that our redirect will work as expected, based on the firebase config)?

petebacondarwin commented:
I am going to try to move this PR forward @alxhub - but I might ping you if I get stuck.

alxhub commented on aio/firebase.json:
```diff
@@ -48,6 +48,11 @@
       {"type": 301, "source": "/guide/service-worker-configref", "destination": "/guide/service-worker-config"}
     ],
     "rewrites": [
+      // Previous Angular service worker redirects to the new script (this must remain forever).
```
I don't believe so - you can't redirect SW scripts as far as I know.

alxhub commented on aio/package.json:
```diff
@@ -76,7 +76,7 @@
     "@angular/platform-server": "^5.1.0-beta.2",
     "@angular/router": "^5.1.0-beta.2",
-    "@angular/service-worker": "^1.0.0-beta.16",
+    "@angular/service-worker": "5.1.0",
```
Yes.

alxhub commented on aio/src/app/documents/document.service.ts:
```diff
@@ -4,6 +4,7 @@ import { HttpClient, HttpErrorResponse } from '@angular/common/http';
 import { Observable } from 'rxjs/Observable';
 import { AsyncSubject } from 'rxjs/AsyncSubject';
 import { of } from 'rxjs/observable/of';
+import 'rxjs/add/observable/of';
```
Indeed!

alxhub commented on aio/src/app/sw-updates/sw-updates.service.ts:
```diff
@@ -51,18 +56,12 @@ export class SwUpdatesService implements OnDestroy {
-        .concat(Observable.of(false)).take(1)
-        .do(v => this.log(`Update available: ${v}`))
-        .subscribe(v => v ? this.activateUpdate() : this.scheduleCheckForUpdate());
+    this.sw.checkForUpdate().then(() => this.scheduleCheckForUpdate());
```
I agree.

alxhub commented on aio/src/app/sw-updates/sw-updates.service.spec.ts:
```diff
@@ -145,16 +137,20 @@ class MockLogger {
+  activated = this.$$activatedSubj.asObservable();
 
-  updates = this.$$updatesSubj.asObservable();
+  get isEnabled(): boolean {
```
No, it probably doesn't need to be.

alxhub commented:
> Thinking about this again, does this mean that files that are not included in the assets will be forwarded to the server (which means that our redirect will work as expected, based on the firebase config)?

No, the SW distinguishes between requests for routes and for files. It can tell routes apart based on 1) `navigate` mode for the request indicating it's a top-level request (navigating to a new page) and 2) absence of a file extension.

mary-poppins commented:
You can preview 1a6ffcc at https://pr19795-1a6ffcc.ngbuilds.io/.

mary-poppins commented:
You can preview 4a1f14a at https://pr19795-4a1f14a.ngbuilds.io/.

mary-poppins commented:
You can preview 572dcd0 at https://pr19795-572dcd0.ngbuilds.io/.

mary-poppins commented:
You can preview 41568fc at https://pr19795-41568fc.ngbuilds.io/.

gkalpak commented:
The problems I currently see:

1. The new SW doesn't seem to be able to handle new clients when offline.

2. Not all redirects will work. More specifically, URLs that do not contain a file extension, will be rewritten to `index.html` by the SW (e.g. `/foo/bar.zip` will work, but `/guide/old-name` won't).
  _We need to either add support for ignoring files to `@angular/service-worker` or implement the redirection on the client._
  _(Concerns: Would SEO be affected? Would need to have a redirection map on the client.)_

3. Minor: For some reason, the `/assets/images/favicons/favicon-YxY.png` files (part of the `assets-lazy` asset group) seem to be fetched on every page transition (and fetched from the server, not the SW cache 😕).

mary-poppins commented:
You can preview 3ad088d at https://pr19795-3ad088d.ngbuilds.io/.

petebacondarwin commented:
What is the status of this PR @alxhub and @gkalpak? It is marked as review but looks more like WIP to me...

gkalpak commented:
MArked is as blocked. We need to (a) investigate why offline doesn't work (and fix it in `@angular/service-worker`) and (b) figure out what to do with the redirects (probably implement the feature in `@angular/service-worker`).

gkalpak commented:
#22883 should fix the offline issue.
#20404 is about adding support for ignoring specific URLs.

gkalpak commented:
The `SwUpdates` test failures should be fixed by https://github.com/angular/zone.js/pull/1057.

gkalpak commented:
BTW, cli@6.0.0-rc.3/4 does not work correctly for SW: https://github.com/angular/angular-cli/issues/10297#issuecomment-381087872

mary-poppins commented:
You can preview 1f280f0 at https://pr19795-1f280f0.ngbuilds.io/.

gkalpak committed:
build(docs-infra): remove unused dependencies

gkalpak committed:
build(docs-infra): enable ServiceWorker in cli config

gkalpak committed:
build(docs-infra): remove the dependency on `rxjs-compat`

mary-poppins commented:
You can preview 771c9e6 at https://pr19795-771c9e6.ngbuilds.io/.

gkalpak commented:
Now blocked on #23526.

gkalpak committed:
refactor(docs-infra): remove unused images

mary-poppins commented:
You can preview 8e9d056 at https://pr19795-8e9d056.ngbuilds.io/.

gkalpak committed:
build(docs-infra): upgrade @angular/cli to 6.0.3

gkalpak committed:
build(docs-infra): upgrade @angular/* to 6.0.2 and rxjs to 6.1.0

gkalpak committed:
build(docs-infra): upgrade @angular/material to 6.0.2

mary-poppins commented:
You can preview 97c34d3 at https://pr19795-97c34d3.ngbuilds.io/.

mary-poppins commented:
You can preview 7e7f224 at https://pr19795-7e7f224.ngbuilds.io/.

gkalpak committed:
refactor(docs-infra): move concept icons to more appropriate location

gkalpak committed:
fix(docs-infra): correctly show icon for fetch error when offline

gkalpak committed:
fix(docs-infra): ignore server-side redirected URLs in the SW

This allows URLs to be passed through to the server (where they are
properly redirected), instead of serving `index.html` from the SW.

Known issue:
`/docs/` will be passed through to the server. `/docs` (without the
trailing slash) will be correctly treated as a navigation URL and
handled by the SW.
We don't link to `/docs/` from within the app, but if there are external
links to `/docs/` they will require a round-trip to the server and will
not work in offline mode.

mary-poppins commented:
You can preview 039be39 at https://pr19795-039be39.ngbuilds.io/.

gkalpak commented:
Blocked on:
- [ ] #24105.
- [ ] `docs-lazy` asset-group URLs being loaded eagerly (despite `installMode` being `lazy`).

gkalpak committed:
test(docs-infra): correctly extract sitemap URLs

`%%DEPLOYMENT_HOST%%` has been assumed to be the host prefix for sitemap
URLs since bf29936af, but afaict this was never the case.

mary-poppins commented:
You can preview 5d9b948 at https://pr19795-5d9b948.ngbuilds.io/.

gkalpak commented on aio/firebase.json:
```diff
@@ -106,6 +106,11 @@
+      // Previous Angular service worker redirects to the new script (this must remain forever).
+      {
+        "source": "/worker-basic.min.js",
+        "destination": "/ngsw-worker.js"
```
@alxhub, now that the new `@angular/service-worker` includes a modified [worker-basic.min.js][1] that just unregisters itself, isn't this rewrite rule unnecessary?

[1]: https://pr19795-ac089ef.ngbuilds.io/worker-basic.min.js

gkalpak committed:
build(docs-infra): update size limits

mary-poppins commented:
You can preview 38e8bb8 at https://pr19795-38e8bb8.ngbuilds.io/.

gkalpak committed:
build(docs-infra): upgrade @angular/cli to 6.1.0-rc.3 and @angular-devkit/build-angular to 0.7.0-rc.3

gkalpak committed:
build(docs-infra): upgrade @angular/* to 6.1.0-rc.3 and rxjs to 6.2.2

gkalpak committed:
build(docs-infra): update size limits

gkalpak committed:
build(docs-infra): downgrade @angular-devkit/build-angular to 0.6.7

The is a bug in versions 0.6.8+ that breaks when trying to use Jasmine's
mock clock.

Related to angular/angular-cli#11626.

gkalpak committed:
fixup! feat(docs-infra): Convert AIO to use the new Service Worker 5.0.0.

mary-poppins commented:
You can preview b924be4 at https://pr19795-b924be4.ngbuilds.io/.

IgorMinar commented on integration/cli-hello-world/tsconfig.json:
```diff
@@ -8,6 +8,7 @@
     "emitDecoratorMetadata": true,
     "experimentalDecorators": true,
     "target": "es5",
+    "skipLibCheck": true,
```
why this change? is it just for speed things up or is there a reason?

gkalpak replied:
I have no idea (or don't remember any more). It is definitely not to speed things up.
I'll try removing it and see if anything breaks.

IgorMinar commented on aio/angular.json:
```diff
@@ -33,7 +33,6 @@
               "src/assets",
               "src/generated",
               "src/app/search/search-worker.js",
-              "src/favicon.ico",
```
how come we no longer need to deal with favicon.ico?

gkalpak replied:
We haven't been using this favicon for a while (so I [deleted it][1]).
We use [another favicon][2] instead. 

[1]: https://github.com/angular/angular/pull/19795/files#diff-bf76692de18e7c4243ffa0f31634419d
[2]: https://github.com/angular/angular/blob/7960d1879d87f1d9e4f7f202423960d874a6a08f/aio/src/assets/images/favicons/favicon.ico

IgorMinar reviewed (approved):
I left a few questions on the PR, but overall this looks great. Sorry for dragging my feet on the review.

Given the risk to production, I think we should be careful about how we roll this out and first test it on next.angular.io.

Here is my proposal:

1. let's merge this into master branch only
2. test on next.angular.io for three scenarios:
  - a browser with preexisting service worker
  - a browser with sw support but no cache or old sw installed
  - a browser without sw support (IE)
3. if all is good, let's wait for a few updates to next.angular.io over the next few days and ensure that the sw is properly updating the caches
4. if all good, let's cherry-pick this PR into the patch branch and deploy to production next week.

gkalpak commented on aio/package.json:
```diff
@@ -56,50 +56,42 @@
-    "ng-pwa-tools": "^0.0.10",
-    "rxjs": "6.0.0-uncanny-rc.7",
-    "rxjs-compat": "6.0.0-uncanny-rc.7",
+    "rxjs": "6.2.2",
```
I'm dropping it in this PR :grin:

gkalpak committed:
fixup! feat(docs-infra): Convert AIO to use the new Service Worker 5.0.0.

gkalpak commented:
The release plan sgtm. Rebased on master and addressed (or responded to) comments.

mary-poppins commented:
You can preview a6c3b62 at https://pr19795-a6c3b62.ngbuilds.io/.

IgorMinar commented:
sgtm. please resolve the conflict and mark this for merge

mary-poppins commented:
You can preview 80382a0 at https://pr19795-80382a0.ngbuilds.io/.

mary-poppins commented:
You can preview 1d82ed3 at https://pr19795-1d82ed3.ngbuilds.io/.

mary-poppins commented:
You can preview 1896fa8 at https://pr19795-1896fa8.ngbuilds.io/.

gkalpak commented:
Note to caretaker:
The `integration_test` CircleCI failure is "inherited" from master and is not related to this PR. (This PR only changes `aio/` stuff and does not affect integration tests.

mary-poppins commented:
You can preview 0832d58 at https://pr19795-0832d58.ngbuilds.io/.
